packages/sitecore-jss/src/personalization/layout-personalization-utils.ts

```diff
@@ -0,0 +1,109 @@
+                var personalizedFragment = personalizedFragments[component.uid];
+                if (personalizedFragment === null) {
+                  hiddenComponents.push(component);
+                } else if (personalizedFragment) {
+                  var wasReplaced = replacePersonalizedRenderings(personalizedFragment);
+                  if (!wasReplaced) {
+                    return false;
+                  }
+                  placeholder[i] = personalizedFragment;
+                } else {
+                  return false;
+                }
+              }
+            }
+            if (hiddenComponents.length) {
+              context.placeholders[key] = placeholder.filter(function (item: any) {
+                return hiddenComponents.indexOf(item) == -1;
+              });
+            }
+          }
+        }
+      }
+      return true;
+    }
+
+    if (replacePersonalizedRenderings(personalizedFragment)) {
+      return personalizedFragment;
+    } else {
+      return defaultComponent;
+    }
+  }
+
+  getPersonalizedComponents(item: any): PersonalizedComponentRendering[] {
+    const result: PersonalizedComponentRendering[] = [];
+    if (isPersonalizedComponentRendering(item)) {
+      result.push(item);
+    } else if (item.placeholders) {
+      for (const key in item.placeholders) {
+        if (item.placeholders.hasOwnProperty(key)) {
+          const placeholder = item.placeholders[key];
+          for (const component of placeholder) {
+            const components = this.getPersonalizedComponents(component);
+            for (const component of components) {
+              result.push(component);
+            }
+          }
+        }
+      }
+    }
+    return result;
+  };
+
+  replacePersonalizedComponentsWithLoaderComponents(placeholders: PlaceholdersData, loaderComponentName: string) {
+    if (placeholders) {
+      for (const key in placeholders) {
+        const placeholder = placeholders[key];
+        if (Array.isArray(placeholder)) {
+          placeholder.forEach((component, index) => {
+            if (hasPersonalization(component) && component.personalization) {
+              const personalizedComponent: PersonalizedComponentRendering = {
+                componentName: loaderComponentName,
+                uid: component.uid as string,
+                personalization: {
+                  hiddenByDefault: component.personalization.hiddenByDefault,
+                  defaultComponent: component.personalization.hiddenByDefault ? null : component
+                }
+              };
+              placeholder[index] = personalizedComponent;
+            } else if (isComponentRendering(component) && component.placeholders) {
+              this.replacePersonalizedComponentsWithLoaderComponents(component.placeholders, loaderComponentName);
+            }
+          });
+        }
+      }
+    }
+  }
+}
+
+export function requirePersonalization(component: any): component is { uid: string } {
```

Comment:
`requirePersonalization` function is only used inside `LayoutPersonalizationUtils`.
Should it be transformed into a private function?